sdk/include/opentelemetry/sdk/trace/tracer_context.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include "opentelemetry/sdk/common/atomic_unique_ptr.h"
+#include "opentelemetry/sdk/resource/resource.h"
+#include "opentelemetry/sdk/trace/processor.h"
+#include "opentelemetry/sdk/trace/samplers/always_on.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+
+/**
+ * A class which stores the TracerProvider context.
+ *
+ * This class meets the following design criteria:
+ * - A shared reference between TracerProvider and Tracers instantiated.
+ * - A thread-safe class that allows updating/altering processor/exporter pipelines
+ *   and sampling config.
+ * - The owner/destroyer of Processors/Exporters.  These will remain active until
+ *   this class is destroyed.  I.e. Sampling, Exporting, flushing etc. are all ok if this
+ *   object is alive, and they will work together. If this object is destroyed, then
+ *   no shared references to Processor, Exporter, Recordable etc. should exist, and all
+ *   associated pipelines will have been flushed.
+ */
+class TracerContext
+{
+public:
+  explicit TracerContext(std::unique_ptr<SpanProcessor> processor,
+                         opentelemetry::sdk::resource::Resource resource =
+                             opentelemetry::sdk::resource::Resource::Create({}),
+                         std::unique_ptr<Sampler> sampler =
+                             std::unique_ptr<AlwaysOnSampler>(new AlwaysOnSampler)) noexcept;
+  /**
+   * Attaches a span processor to this tracer context.
+   *
+   * @param processor The new span processor for this tracer. This must not be
+   * a nullptr.  Ownership is given to the `TracerContext`.
+   */
+  void RegisterPipeline(std::unique_ptr<SpanProcessor> processor) noexcept;
+
+  /**
+   * Obtain the sampler associated with this tracer.
+   * @return The sampler for this tracer.
+   */
+  Sampler &GetSampler() const noexcept;
+
+  /**
+   * Obtain the (conceptual) active processor.
+   *
+   * Note: When more than one processor is active, this will
+   * return an "aggregate" processor
+   */
+  SpanProcessor &GetActiveProcessor() const noexcept;
```

Comment:
We might theoretically run into issue here when some tracer does some work with a processor obtained by `GetActiveProcessor`, while another thread calls `RegisterPipeline` and thus causes the memory pointed to by the obtained the processor to be freed.

Reply:
Shouldn't be the case.  The interface is "RegisterPipeline", and it's a shared pointer.  The TracerContext can guarantee that ALL registered processors remain alive for its own lifecycle.   I.e. we're not going to delete a pipeline, we're only going to allow registration of new ones specifically to prevent this issue.

We have the inverse problem around Recordable though, specifically if we have an "active" span that did NOT get a recordable registered when a new pipeline is registered, what happens when that span ends?

If you look at my design around "ExportableSpan" where a pipeline can ask for its instance of Recordable (and handle a missing recordable), that's how I planned to solve that issue.

Reply:
I think I understand. So if a new pipeline is registered, the old one isn't just overwritten, but kept in store and just made _inactive_?

Reply:
My understanding was that all the pipelines(i.e, processors) remain active. And we can only add a new pipeline, but can't remove the existing one? And once we add multi-processor support, `TracerContext` would actually maintain a unique_ptr to (say) `CompositeSpanProcessor` which further manages all the active pipelines/processors?

Reply:
Yeah.  i'm adopting the design I've seen elsewhere where you can "append a pipeline", but can't remove an existing pipeline.

I.e. once a pipeline is active, it's forever active.   We don't have a register/unregister mechanism.